File: src/chipschallenge/clonebehaviors/BlobCloneBehavior.java

package chipschallenge.clonebehaviors;

import chipschallenge.Block;


public class BlobCloneBehavior implements CloneBehavior {

    private BlobCloneBehavior() {
    }
    private static BlobCloneBehavior mInstance = null;

    public static synchronized BlobCloneBehavior getInstance() {
        if (mInstance == null) {
            mInstance = new BlobCloneBehavior();
        }
        return mInstance;
    }

    public Block cloneIt(Block original) throws CloneNotSupportedException {
        throw new CloneNotSupportedException();
    }
}

File: ChipsChallenge.Shared/Clonebehaviors/BlobCloneBehavior.cs
namespace ChipsChallenge.Shared.Clonebehaviors
{
    using Moves = Move.Moves;

    public class BlobCloneBehavior : CloneBehavior
    {
        private static BlobCloneBehavior instance;

        private BlobCloneBehavior()
        {
        }

        public static BlobCloneBehavior Instance
        {
            get
            {
                lock (typeof(BlobCloneBehavior))
                {
                    return instance ?? (instance = new BlobCloneBehavior());
                }
            }
        }

        public override Block CloneIt(Block original)
        {
            int dir = Utils.Random.Next(4);
            Block clone = null;
            for (int i = 0; i < 2; i++)
            {
                switch (dir)
                {
                    case 0:
                        if ((clone = CloneTo(original, Moves.UP)) != null)
                        {
                            goto outerBreak;
                        }
                        break;
                    case 1:
                        if ((clone = CloneTo(original, Moves.DOWN)) != null)
                        {
                            goto outerBreak;
                        }
                        break;
                    case 2:
                        if ((clone = CloneTo(original, Moves.LEFT)) != null)
                        {
                            goto outerBreak;
                        }
                        break;
                    case 3:
                        if ((clone = CloneTo(original, Moves.RIGHT)) != null)
                        {
                            goto outerBreak;
                        }
                        else
                        {
                            dir = 0;
                        }
                    break;
                }
            }
            outerBreak:
            return clone;
        }
    }
}